homeassistant/components/speedtestdotnet/__init__.py

```diff
@@ -3,83 +3,132 @@
     )
 
     return True
 
 
-class SpeedtestData:
+async def async_unload_entry(hass, config_entry):
+    """Unload SpeedTest Entry from config_entry."""
+    hass.services.async_remove(DOMAIN, SPEED_TEST_SERVICE)
+
+    await hass.config_entries.async_forward_entry_unload(config_entry, "sensor")
+
+    hass.data.pop(DOMAIN)
+
+    return True
+
+
+class SpeedTestDataCoordinator(DataUpdateCoordinator):
     """Get the latest data from speedtest.net."""
 
-    def __init__(self, hass, server_id):
+    def __init__(self, hass, config_entry):
         """Initialize the data object."""
-        self.data = None
-        self._hass = hass
-        self._servers = [] if server_id is None else [server_id]
+        self.hass = hass
+        self.config_entry = config_entry
+        self.api = None
+        self.servers = {}
+        super().__init__(
+            self.hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_method=self.async_update,
+            update_interval=timedelta(
+                seconds=self.config_entry.options.get(
+                    CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+                )
+            ),
+        )
 
-    def update(self, now=None):
+    def update_data(self):
         """Get the latest data from speedtest.net."""
+        server_list = self.api.get_servers()
+
+        self.servers[DEFAULT_SERVER] = None
+        for server in sorted(
+            server_list.values(), key=lambda server: server[0]["country"]
+        ):
+            self.servers[f"{server[0]['country']} - {server[0]['name']}"] = server
+
+        if self.config_entry.options.get(CONF_SERVER_ID):
+            server_id = self.config_entry.options.get(CONF_SERVER_ID)
+            self.api.closest.clear()
+            self.api.get_servers(servers=[server_id])
+            self.api.get_best_server()
+        _LOGGER.debug(
+            "Executing speedtest.net speed test with server_id: %s", self.api.best["id"]
+        )
+
+        self.api.download()
+        self.api.upload()
+        return self.api.results.dict()
+
+    async def async_update(self, *_):
+        """Update Speedtest data."""
+        try:
+            return await self.hass.async_add_executor_job(self.update_data)
+        except (speedtest.ConfigRetrievalError, speedtest.NoMatchedServers):
+            raise UpdateFailed
 
-        _LOGGER.debug("Executing speedtest.net speed test")
-        speed = speedtest.Speedtest()
-        speed.get_servers(self._servers)
-        speed.get_best_server()
-        speed.download()
-        speed.upload()
-        self.data = speed.results.dict()
-        dispatcher_send(self._hass, DATA_UPDATED)
+    async def async_setup(self):
+        """Set up SpeedTest."""
+        try:
+            self.api = await self.hass.async_add_executor_job(speedtest.Speedtest)
+        except speedtest.ConfigRetrievalError:
+            raise ConfigEntryNotReady
+
+        async def request_update(event):
```

Comment:
The argument passed to the service handler is a `ServiceCall` instance.
```suggestion
        async def request_update(call):
```